File: wallet-web/app/src/api/assetsApi.ts
import { apiRequest } from './httpClient';
import { assertOnline } from '../connectivity/assertOnline';
import type { ActionProposal } from './paymentsApi';

export interface CuratedToken {
  id: number;
  symbol: string;
  contractAddress: string;
  decimals: number;
  name: string;
  imageUrl: string;
  isActive: boolean;
}

export function listCuratedTokens(): Promise<CuratedToken[]> {
  return apiRequest<CuratedToken[]>('/v1/assets');
}

/** tokenAddress omitted/empty = native ETH balance (wallet-backend's
 * GET /v1/assets/balance/:address?token=... - PLAN.md §2). */
export function getBalance(address: string, tokenAddress?: string): Promise<string> {
  const query = tokenAddress ? `?token=${encodeURIComponent(tokenAddress)}` : '';
  return apiRequest<{ balance: string }>(`/v1/assets/balance/${address}${query}`).then((r) => r.balance);
}

export async function buildApprove(
  walletAddress: string,
  tokenAddress: string,
  spender: string,
  amount: string,
): Promise<ActionProposal> {
  await assertOnline();
  return apiRequest<ActionProposal>('/v1/assets/approve/build', {
    method: 'POST',
    walletAddress,
    body: { tokenAddress, spender, amount },
  });
}

export function submitApprove(walletAddress: string, actionId: number, signature: string): Promise<{ hash: string }> {
  return apiRequest<{ hash: string }>('/v1/assets/approve/submit', {
    method: 'POST',
    walletAddress,
    body: { actionId, signature },
  });
}

File: wallet-mobile/app/lib/api/assets_api.dart
// Mirrors wallet-web's api/assetsApi.ts.
import 'http_client.dart';

class CuratedToken {
  CuratedToken({required this.id, required this.symbol, required this.contractAddress, required this.decimals, required this.name});

  final int id;
  final String symbol;
  final String contractAddress;
  final int decimals;
  final String name;

  factory CuratedToken.fromJson(Map<String, dynamic> json) => CuratedToken(
        id: json['id'] as int,
        symbol: json['symbol'] as String,
        contractAddress: json['contractAddress'] as String,
        decimals: json['decimals'] as int,
        name: json['name'] as String,
      );
}

class AssetsApi {
  AssetsApi(this._client);
  final ApiClient _client;

  /// GET /v1/assets (public).
  Future<List<CuratedToken>> listCuratedTokens() {
    return _client.request(
      '/v1/assets',
      decode: (json) => (json as List).map((e) => CuratedToken.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }

  /// GET /v1/assets/balance/:address?token=... (public). Omitted/empty
  /// tokenAddress = native ETH balance.
  Future<String> getBalance(String address, {String? tokenAddress}) {
    final query = tokenAddress != null ? '?token=${Uri.encodeComponent(tokenAddress)}' : '';
    return _client.request(
      '/v1/assets/balance/$address$query',
      decode: (json) => (json as Map<String, dynamic>)['balance'] as String,
    );
  }
}
